Mainnet/HashBattle/HashBattle/Arena.cs

```diff
@@ -0,0 +1,194 @@
+﻿using Stratis.SmartContracts;
+using System;
+using System.Text;
+
+/// <summary>
+/// A Stratis smart contract for running a game battle where owner will start the battle and maximum 4 users can enter a battle
+/// </summary>
+public class Arena : SmartContract
+{
+    public Arena(ISmartContractState smartContractState)
+    : base(smartContractState)
+    {
+        BattleOwner = Message.Sender;
+    }
+
+    /// <summary>
+    /// Set the address deploying the contract as battle owner
+    /// </summary>
+    private Address BattleOwner
+    {
+        get => PersistentState.GetAddress(nameof(BattleOwner));
+        set => PersistentState.SetAddress(nameof(BattleOwner), value);
+    }
+
+    /// <summary>
+    /// Battle owner will start the battle
+    /// </summary>
+    public bool StartBattle(ulong battleId, ulong fee)
+    {
+        Assert(Message.Sender == BattleOwner, "Only battle owner can start game.");
+
+        var battle = new BattleMain();
+        battle.BattleId = battleId;
+        battle.MaxUsers = 4;
+        battle.Fee = fee;
+        battle.Users = new Address[battle.MaxUsers];
+        SetBattle(battleId, battle);
+
+        Log(battle);
+        return true;
+    }
+
+    /// <summary>
+    /// 4 different user will enter the battle
+    /// </summary>
+    public bool EnterBattle(ulong battleId, uint userindex)
+    {
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle ended.");
+
+        Assert(battle.Fee == Message.Value, "Battle amount is not matching.");
+
+        var user = GetUser(battleId, Message.Sender);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Address = Message.Sender;
+
+        SetUser(battleId, Message.Sender, user);
+
+        battle.Users.SetValue(user.Address, userindex);
+        SetBattle(battleId, battle);
+
+        Log(battle);
```

Comment:
You may want to consider logging events rather than the battle details. For example, you could define an event like so:

```csharp
public struct HashBattleEnteredEvent
{
    public ulong BattleId;
    public Address Address;
}
```

Then here you would call

```csharp
Log(new HashBattleEnteredEvent { BattleId = battleId, Address = Message.Sender }
```

This way it'd be easier to index and process the logs on your dApp. You're able to search the full node for specific events rather than having to analyze the entire sequence of events to determine which action happened at a point in time.